### PR TITLE
Make F# resources work more like C#

### PR DESCRIPTION
**Pull request by nosami**
Given a resource in a folder named `AFolder/AResourceFile.txt` in a project with
a root namespace of `fsharpresources`, then the following command line arg
should be passed to the compiler

```
--resource:AFolder/AResourceFile.txt,fsharpresources.AFolder.AResourceFile.txt
```

whereas currently it just passes the name of the file. This also fixes passing
a `LogicalName` to the resource.

Fixes https://github.com/Microsoft/visualfsharp/issues/1050 &&
https://github.com/Microsoft/visualfsharp/issues/922

I looked at how Roslyn was doing this [here](https://github.com/dotnet/roslyn/blob/6847f1e5a909395aae9456e8f366cbf4deb86b69/src/Compilers/Core/MSBuildTask/ManagedCompiler.cs#L739)

### CODE REVIEW

**Comment by nosami** [comment]
This also fixes https://bugzilla.xamarin.com/show_bug.cgi?id=57689 and https://bugzilla.xamarin.com/show_bug.cgi?id=37971

**Comment by dsyme** [comment]
@nosami This has been a huge source of pain historically - really glad you have sorted this out

Could you also submit to fsharp/fsharp and we'll check what the extra testing there says (there is testing enabled in that repo that is not enabled here)

**Comment by nosami** [comment]
It took me most of today to work that out :)

Will submit to fsharp/fsharp... I tested with that repo locally, then applied the PR here :)

**Comment by KevinRansom** [comment]
@dsyme what is the nature of the extra testing?

**Comment by nosami** [comment]
Fwiw, I ran the Xamarin tests that build each F# template in VS for Mac via msbuild with this change applied. All passed.

**Comment by dsyme** [comment]
@KevinRansom I think it is
* [the compilation of tests/projects against an installation of F# on Mono](https://github.com/fsharp/fsharp/blob/master/mono/test-mono.sh#L5)
* a matrix of Mono installations (nightly, weekly, specific versions) automated by Travis https://github.com/fsharp/fsharp/blob/master/.travis.yml
* OSX testing https://github.com/fsharp/fsharp/blob/master/.travis.yml

The problems in this repo are
* all we do here is build. https://github.com/Microsoft/visualfsharp/blob/master/netci.groovy#L57. 
* I'm also unhappy with the massive duplication in build.sh/build.cmd in this repo, which is clearly unsustainable.

For now I've personally run out of time w.r.t. on the repo unification work and have to focus on language design and research work.

**Comment by dsyme** [comment]
>  Fwiw, I ran the Xamarin tests that build each F# template in VS for Mac via msbuild with this change applied. All passed.

@nosami Great to know you have these tests.  Resources are used nearly always in UI.  Do the tests also run the resulting apps?

**Comment by KevinRansom** [comment]
@Don Syme<mailto:Don.Syme@microsoft.com>

You aint kidding about the Cambridge and QA test suites.  But nothing gets achieved if nothing gets done on them.

**Comment by nosami** [comment]
@dsyme The tests just download nuget packages, check that msbuild works, and that there are no squiggles in the editor - https://github.com/mono/monodevelop/blob/master/main/external/fsharpbinding/MonoDevelop.FSharp.Tests/TemplateTests.fs

**Comment by dsyme** [comment]
@nosami I'm concerned that the generated names of resources may be wrong - this usually shows up as a runtime failure in some "GetResource" call

**Comment by dsyme** [comment]
@nosami You have one failure: 

```
1) Failed : Tests.ProjectSystem.Miscellaneous.CreateFSharpManifestResourceName
[("Abc.resources", "Abc.resources"); ("Bar.de.resx", "Bar.de");
 ("Bar.resx", "Bar"); ("Xyz\Baz.ru.resx", "Xyz.Baz.ru")]<>[("Abc.resources", "RootNamespace.Abc.resources");
 ("Bar.de.resx", "RootNamespace.Bar.de"); ("Bar.resx", "RootNamespace.Bar");
 ("Xyz\Baz.ru.resx", "RootNamespace.Xyz.Baz.ru")]
at <StartupCode$VisualFSharp-Unittests>.$Tests.ProjectSystem.Miscellaneous.CreateFSharpManifestResourceName@444.Invoke(String file) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.ProjectSystem.Miscellaneous.fs:line 454
at UnitTests.TestLib.Utils.FilesystemHelpers.DoWithTempFile[a](String filename, FSharpFunc`2 f) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\TestLib.Utils.fs:line 73
at Tests.ProjectSystem.Miscellaneous.CreateFSharpManifestResourceName() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.ProjectSystem.Miscellaneous.fs:line 444

```

**Comment by nosami** [comment]
Well, the generated names may well be different now, this is definitely a breaking change. But, I think it's a good breaking change - it looks as if it's never worked correctly.

I think that the test is wrong and needs to be changed to match the code. The new names look correct to me.

**Comment by nosami** [comment]
As far as mono goes, this change is needed. Some things in Xamarin iOS just plain don't work https://bugzilla.xamarin.com/show_bug.cgi?id=57689 being a prime example. I would rather have that work and take the fallout from the change.

**Comment by dsyme** [comment]
>  As far as mono goes, this change is needed. Some things in Xamarin iOS just plain don't work https://bugzilla.xamarin.com/show_bug.cgi?id=57689 being a prime example. I would rather have that work and take the fallout from the change.

Yes, I'm inclined to believe that.  I don't have access to that specific bug though

@cartermp and @KevinRansom will need to make the call about the breaking change.  Are the scenarios simply 
1. "I have resources in my project and they are in a sub-folder"?  
2. "I have a RootNamespace element in my project file and use some resources"?

Is there anything we could do to moderate the breaking change?  e.g. only activate it if some other setting is passed in, and set that in the Xamarin projects and all future templates?

**Comment by nosami** [comment]
The main issue is that you can't have the same filename in multiple folders. This happens frequently for Xamarin, e.g. the same icon but in different sizes - they get put in subfolders but the filename is the same. 

It would be far too difficult to try and work against that with F# specific hacks.

**Comment by dsyme** [comment]
> It would be far too difficult to try and work against that with F# specific hacks.

I totally agree, I want to see this fixed once-and-for-all.  

But can we make this conditionally activated in some way by a ``<UseCorrectResourceNames>true</UseCorrectResourceNames>`` setting or something?  Or even by using new-style project files?  Or some Xamarin project file setting?  We just need to make sure this is not breaking for existing projects.

**Comment by matthid** [comment]
is this actually breaking? Was using folders ever officially supported? It definitely never felt like it was.

**Comment by dsyme** [comment]
> is this actually breaking? Was using folders ever officially supported? It definitely never felt like it was.

@matthid Yes, I believe you could do it and inspect the assembly for the resource names generated.  Also this now respects ``<RootNamespace>Foo</RootNamespace>`` which was previously ignored

**Comment by nosami** [comment]
> Yes, I'm inclined to believe that. I don't have access to that specific bug though

Do you have a bugzilla account? I thought this was a public bug

As this is quite risky (even though I feel it's right), I could apply this patch to mono only. That way only mono / xamarin users would be affected... but then Windows users would still be affected

I feel like that the correct solution would be to have a switch that reverted this behaviour.

**Comment by dsyme** [comment]
> As this is quite risky (even though I feel it's right), I could apply this patch to mono only. That way only mono / xamarin users would be affected... but then Windows users would still be affected

I think we want this fix to apply for Xamarin-on-Visual-Studio-for-Windows too.  Inconsistency in Xamarin tooling between Windows and OSX and Linux would be bad

>  I feel like that the correct solution would be to have a switch that reverted this behaviour.

Are you sure we can't activate it explicitly?  That would be the safest route. We don't want existing F# code breaking.

**Comment by KevinRansom** [comment]
@dsyme 

It's interesting that this is not a compiler fix, but instead a fix in the build task modification just changing what's passed to the compiler.

It seems that we could and should enable this as the default approach.  And add a buildtask option to use the old behavior to enable as a quick fix for anyone whose app breaks.

I don't think there is a lot of risk of making this change, if we have a simple fallback.

Kevin

**Comment by nosami** [comment]
> Are you sure we can't activate it explicitly? That would be the safest route. We don't want existing F# code breaking.

For sure, we can do this. But it feels (to me) like adding a switch to make it behave like it should behave is wrong - there should be a switch to enable the opposite.

**Comment by dsyme** [comment]
@nosami  Could you also review [these notes](https://fsharp.github.io/2015/09/29/fsharp-compiler-guide.html#how-f-handles-embeddedresource-items-differently-to-c-details) please?    This is where I documented all the variations in behaviour once - I didn't know what or how to fix the problem at root cause

Perhaps write up a new version of all those scenarios too.

> I don't think there is a lot of risk of making this change, if we have a simple fallback.

This is not right. Do we really want upgrading to VS2017 15.4 or the latest Xamarin package to just give random "missing resource" failures on startup of applications? 

Silently changing the names of generated resources has busted, among other things, the compiler packages and apps  on multiple occasions in the past.  
* For the compiler package the resources suddenly had wrong embedded names. 
* For the Xamarin/MonoDevelop addin the names were inconsistent on Windows and Mono, so it depended where you compiled things.

In both cases the failures only manifest at runtime.  

I absolutely want this problem to cease, but I'm 100% convinced that this is a risky breaking change to compilations that use resources - including the two cases where I have used them - since most people end up putting resources in directories sooner or later.  We would only consider this safe if we were 99.9% sure that no one was doing that - but we know that they are. We were nervous about fixing this blindly when we first came across the whacky behaviour (and unimplemented RootNamespace behaviour) in F# 3.0 timeframe.  If we were nervous then, we should be even more careful now.

I really want this fixed, but we should make it opt-in and non-breaking for existing project files (it can be the default for new .NET SDK project files)

**Comment by dsyme** [comment]
@nosami Is the RootNamespace part of the fix necessary here?  My understanding from your comment above is that it isn't?

BTW the steps in [this bug](https://github.com/Microsoft/visualfsharp/issues/922) indicate all that's required for a user to have done for her to be faced with a breaking change here:

1. Create a new F# Library project
2. Add a folder called "AFolder"
3. Add a text file to AFolder called "AResourceFile.txt"
4. Set the build action on that text file to embedded resource

That is a **very** common scenario.  Many people will have embedded resources into their applications or libraries at some point.  They will now get a silent "resource not found" or null pointer exception at runtime if they recompile.

I hate the fact that this was botched in F# 2.0 - and we have to dig ourselves out of the hole. But silently breaking existing code isn't the way out.

**Comment by nosami** [comment]
@dsyme: You know the risks far better than I do :) I know that this is super risky. Whatever you decide is fine with me :)

**Comment by dsyme** [comment]
@nosami  The only thing I can think of is a new setting

    <UseStandardResourceNames>true</UseStandardResourceNames>

(or some similar property) which we put in F# VS and Xamarin templates for old-style project files, and we always activate it for new-style project files.  Any chance you could you draft that?

**Comment by KevinRansom** [comment]
@dsyme 
We will make this the default for SDK projects.  And for new projects created using VS.

**Comment by dsyme** [comment]
> @dsyme We will make this the default for SDK projects. And for new projects created using VS.

Super.

**Comment by 7sharp9** [comment]
Being able to have multi dpi images in F# Android Applications is a must really, with resources being broken for so long its difficulty to recommend  F# for android at all.

**Comment by dsyme** [comment]
>  Being able to have multi dpi images in F# Android Applications is a must really

Yes, it must be fixed

**Comment by nosami** [comment]
OK. I've added a `UseStandardResources` flag which is disabled by default.

With
```
    <UseStandardResourceNames>true</UseStandardResourceNames>
```
we get the following output

```
--resource:AFolder/AResourceFile.txt,fsharpresource.AFolder.AResourceFile.txt
```

With the value omitted or set to false, the behaviour is as it was previously.

```
--resource:AFolder/AResourceFile.txt
```

File: src/fsharp/FSharp.Build/Microsoft.FSharp.Targets [hunk 1]
```diff
@@ -89,24 +89,28 @@ this file.
 
         <CreateFSharpManifestResourceName
              Condition="'@(ResxWithNoCulture)' != '' AND '$(UsingXBuild)' == 'true'"
+             UseStandardResourceNames="$(UseStandardResourceNames)"
```
Comment (dsyme) [inline src/fsharp/FSharp.Build/Microsoft.FSharp.Targets]: @KevinRansom Does anything need to be done for .NET SDK targets here?

**Comment by dsyme** [comment]
@nosami Great.  Could you also adjust that test that was failing to run in both modes and check each?  And check that we have  RootNamespace=null as a variation on the test too?

**Comment by nosami** [comment]
Sure - will do.

**Comment by nosami** [comment]
Writing the tests might not be so easy - I only have OSX on this laptop and VisualFSharp doesn't compile there. I can try and do it "blind"

File: src/fsharp/FSharp.Build/Microsoft.FSharp.Targets (hunk 1 above)
Comment (KevinRansom) [inline src/fsharp/FSharp.Build/Microsoft.FSharp.Targets]: @dsyme. I will update VS templates and the dotnet sdk.

Comment (ErikSchierboom) [reply src/fsharp/FSharp.Build/Microsoft.FSharp.Targets]: @KevinRansom Is this fix in the latest .NET SDK or will it be in a future release? I cannot get it to work on my project, which looks like this:

```xml
<PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <RootNamespace>Generators</RootNamespace>
    <UseStandardResourceNames>true</UseStandardResourceNames>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="Program.fs" />
  </ItemGroup>

  <ItemGroup>
    <EmbeddedResource Include="Templates\*.liquid" />
  </ItemGroup>
```

If I use `ildasm` to look at the manifest, the resource doesn't have the prepended namespace:

```.mresource public _TestClass.liquid```
